src/sio.ts

```diff
@@ -64,129 +70,152 @@ export class RPSIO {
   }
 
   readUint32(offset: number) {
-    switch (offset) {
-      case GPIO_IN:
-        return 0; // TODO implement!
-      case GPIO_HI_IN:
-        return 0; // TODO implement!
-      case GPIO_OUT:
-        return this.gpioValue;
-      case GPIO_OE:
-        return this.gpioOutputEnable;
-      case GPIO_HI_OUT:
-        return this.qspiGpioValue;
-      case GPIO_HI_OE:
-        return this.qspiGpioOutputEnable;
-      case GPIO_OUT_SET:
-      case GPIO_OUT_CLR:
-      case GPIO_OUT_XOR:
-      case GPIO_OE_SET:
-      case GPIO_OE_CLR:
-      case GPIO_OE_XOR:
-      case GPIO_HI_OUT_SET:
-      case GPIO_HI_OUT_CLR:
-      case GPIO_HI_OUT_XOR:
-      case GPIO_HI_OE_SET:
-      case GPIO_HI_OE_CLR:
-      case GPIO_HI_OE_XOR:
-        return 0; // TODO verify with silicone
-      case CPUID:
-        // Returns the current CPU core id (always 0 for now)
+    if (offset < SPINLOCK0) {
+      //not a spinlock
+      switch (offset) {
+        case GPIO_IN:
+          return 0; // TODO implement!
+        case GPIO_HI_IN:
+          return 0; // TODO implement!
+        case GPIO_OUT:
+          return this.gpioValue;
+        case GPIO_OE:
+          return this.gpioOutputEnable;
+        case GPIO_HI_OUT:
+          return this.qspiGpioValue;
+        case GPIO_HI_OE:
+          return this.qspiGpioOutputEnable;
+        case GPIO_OUT_SET:
+        case GPIO_OUT_CLR:
+        case GPIO_OUT_XOR:
+        case GPIO_OE_SET:
+        case GPIO_OE_CLR:
+        case GPIO_OE_XOR:
+        case GPIO_HI_OUT_SET:
+        case GPIO_HI_OUT_CLR:
+        case GPIO_HI_OUT_XOR:
+        case GPIO_HI_OE_SET:
+        case GPIO_HI_OE_CLR:
+        case GPIO_HI_OE_XOR:
+          return 0; // TODO verify with silicone
+        case CPUID:
+          // Returns the current CPU core id (always 0 for now)
+          return 0;
+        case SPINLOCK_ST:
+          return this.spinLock;
+        case DIV_UDIVIDEND:
+          return this.divDividend;
+        case DIV_SDIVIDEND:
+          return this.divDividend;
+        case DIV_UDIVISOR:
+          return this.divDivisor;
+        case DIV_SDIVISOR:
+          return this.divDivisor;
+        case DIV_QUOTIENT:
+          this.divCSR &= ~0b10;
+          return this.divQuotient;
+        case DIV_REMAINDER:
+          return this.divRemainder;
+        case DIV_CSR:
+          return this.divCSR;
+      }
+    }
+    if (offset >= SPINLOCK0 && offset <= SPINLOCK31) {
```

Comment:
I suggest to move this `if` above the `switch`, and then keep the switch as it is. Since this if always returns something, we'll only get to the switch if the offset is not one of the spinlocks...

e.g., this is the structure that I have in mind

```
if (offset >= SPINLOCK0 && offset <= SPINLOCK31) {
   // ...
}

switch (offset) {
   // ...
}

console.warn(`Read from invalid SIO address: ${offset.toString(16)}`);
return 0xffffffff;
```